sdk/storage/azure-storage-data-movement/src/main/java/com/azure/storage/data/movement/PathScanner.java

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.storage.data.movement;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import com.azure.core.util.logging.ClientLogger;
+
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.stream.BaseStream;
+
+public class PathScanner {
+    private final ClientLogger logger = new ClientLogger(PathScanner.class);
+    private final String basePath;
+
+    public PathScanner(String path) {
+        try {
+            // Resolve the given path to an absolute path in case it isn't one already
+            Path pathObj = Paths.get(path);
+            basePath = pathObj.toAbsolutePath().toString();
+
+            // Check if the path exists and whether or not it's a directory; throw
+            // an error if there's nothing present or readable at the given path
+            if (!(Files.isDirectory(pathObj) || Files.isRegularFile(pathObj))) {
+                throw new IllegalArgumentException(String.format("No item(s) located at the path '%s'.", basePath));
+            }
+        } catch (RuntimeException e) {
+            throw logger.logExceptionAsError(e);
+        }
+    }
+
+    public Flux<String> scan(boolean skipSubdirectories) {
+        // Set up Mono for recursing through directories
+        Mono<Path> basePath = Mono.just(Paths.get(this.basePath));
+
+        // Return a Flux constructed from expanding path
+        return basePath.expand(path -> {
+            // Return an empty publisher when the path is not a directory
+            // (files will cause errors when being used with Files::list)
+            if (!Files.isDirectory(path)) {
+                return Mono.empty();
+            } else {
+                return Flux.using(() -> Files.list(path),
```

Comment:
Is this going to do what we want? Looking at `Files.list` it states the listing isn't recursive.

Reply:
Right, `Files.list` itself isn't recursive, but the `Mono.expand` call this is wrapped in has `Files.list` recusively call on its own returned values (provided they're directories; else it won't expand it).